Makefile

```diff
@@ -52,23 +56,49 @@ else
 GOBIN=$(shell go env GOBIN)
 endif
 
+ifeq ($(strip ${PKG_CONFIG_PATH}),)
+	MAKE_PKG_CONFIG_PATH = $(LIBGIT2_LIB_PATH)/pkgconfig
+else
+	MAKE_PKG_CONFIG_PATH = ${PKG_CONFIG_PATH}:$(LIBGIT2_LIB_PATH)/pkgconfig
+endif
+
 all: build
 
 build: $(LIBGIT2) ## Build manager binary
-	PKG_CONFIG_PATH=$(LIBGIT2_LIB_PATH)/pkgconfig/ \
+ifdef MAC_M1
+	PKG_CONFIG_PATH=$(MAKE_PKG_CONFIG_PATH) \
+	CGO_LDFLAGS="-Wl,-rpath,$(LIBGIT2_LIB_PATH)" \
```

Comment:
Is this breaking on macOS Intel? If not I would replace `MAC_M1` with `Darwin`.

Reply:
Don't have an Intel mac to test it on, I'm afraid. This seemed the safer way to go, since the current Makefile works on Intel Macs.

Reply:
@squaremo can you please test this on your mac?

Reply:
My mac may not be the best for testing this, since it's not a newcomer to the libgit2 battlefront. I've got libraries, pkg-config files, and env entries littered everywhere. Indeed, `make run` works with the changes, without the changes, and when I remove everything except for `go run ./main.go`.

I would like to understand why -rpath works where LD_CONFIG_PATH doesn't, since these do pretty similar things. That explanation could usefully go in the commit message :-)

Reply:
I'm not an expert on dynamic linking, but I'll try to explain what I _think_ is going on.

```
> otool -L bin/manager
bin/manager:
        /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1853.0.0)
        /System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 60157.30.13)
        @rpath/libgit2.1.1.dylib (compatibility version 1.1.0, current version 1.1.1)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
 ```
 The thing of our interest is `@rpath/libgit2.1.1.dylib`. From [this](https://www.mikeash.com/pyblog/friday-qa-2009-11-06-linking-and-install-names.html) article about `@rpath`:
 > Starting in 10.5, Apple provides `@rpath`, which is a solution to this. When placed at the front of an install name, this asks the dynamic linker to search a list of locations for the library.
 
 So, at runtime when our binary tries to load `libgit2.1.1.dylib`, it tries to search for it in the dirs that fall under `@rpath`. Presumably, the linker ignores `LD_LIBRARY_PATH` because of the usage of `@rpath`.

 ---
 Why this is happening only on M1 and not on Intel, is beyond my understanding I'm afraid.

Reply:
(After reminding myself how dynamic linking works on MacOS...)

> otool -L bin/manager
> bin/manager:
>         /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1853.0.0)
>         /System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 60157.30.13)
>         @rpath/libgit2.1.1.dylib (compatibility version 1.1.0, current version 1.1.1)
>         /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)

This shows the install name of the library; an `@rpath` at the start says "look for this in the paths embedded in the executable". But what are those embedded paths? That's what the linker option `-rpath` does: if you do

```console
otool -l ./bin/manager # NB lowercase 'l'
```

... in the output, you'll see something like

```
Load command 16                                                                                                                                                                                                                               
          cmd LC_RPATH                                                                                                                                                                                                                        
      cmdsize 72                                                                                                                                                                                                                              
         path /Users/mikeb/work/source-controller/hack/libgit2/lib (offset 12) 
```

This is one of those paths. So: the `@rpath` in the `otool -L` output is something go decides to do; the `LC_RPATH` is embedded when you send `-rpath ...` to the linker. They work together so that the lib is found when you run the executable.

Instead of embedding the library search path in the executable, you can supply it with `DYLD_LIBRARY_PATH` (not `LD_LIBRARY_PATH`, on MacOS -- my mistake). So you can leave out the `-rpath` linker flag, and do

```console
DYLD_LIBRARY_PATH=$PWD/hack/libgit2/lib ./bin/manager
```

and it'll work in the same way. Embedding the path is better though, since then you don't have to remember the env entry when running `./bin/manager`.

The last mystery is: why did it work _without_ setting the library search path (and having the wrong env entry)? I'm not sure it did! Or if it did, only because there was coincidentally a library in the built-in search path, e.g., in `/usr/local/lib`. (Another explanation is that for whatever reason, go was embedding the absolute path to the library, so the search path wasn't needed -- perhaps that's what go 1.17 does, or perhaps that just triggered a rebuild of relevant packages.)

Reply:
I spun up my very old Intel Mac running macOS 10.13.  The old `Makefile` errored out with the same error whereas the new `Makefile` works just the same as it does on macOS arm64. I'm starting to think that it was working on Intel Macs accidentally rather than the way it was intended to. It might be something to do with the go version like you said (i ran go 1.17 on both laptops). Let me try running it using go 1.16 on my Intel Mac and get back to you.

Reply:
> I'm starting to think that it was working on Intel Macs accidentally

I think it is less likely to be to do with the CPU, and more to do with the individual machines on which it gets run; or failing that explanation, the OS version.

Reply:
Same error on macOS 10.13 running go 1.16 :)

Reply:
I think it used to work for us because we've installed libgit2 with brew, back when the version matched what Flux wants.

Reply:
> I think it used to work for us because we've installed libgit2 with brew, back when the version matched what Flux wants.

Yeah I think you're right -- we both did this, and that's pretty much the pool of test machines accounted for :-)

(By the way, I verified that it was working by coincidence by moving /usr/local/lib/libgit2.* out of the way -- the executable then fails to start.)

@aryan9600 I think you can reframe this PR as "Work by design rather than by coincidence, on MacOS" :-) (And you don't need to have a special case for M1 Macs, to answer the original question ...)